src/addrman.cpp

```diff
@@ -714,72 +714,95 @@ void AddrManImpl::Attempt_(const CService& addr, bool fCountFailure, NodeSeconds
-            assert(it_found != mapInfo.end());
-            const AddrInfo& info{it_found->second};
-            // With probability GetChance() * fChanceFactor, return the entry.
-            if (insecure_rand.randbits(30) < fChanceFactor * info.GetChance() * (1 << 30)) {
-                LogPrint(BCLog::ADDRMAN, "Selected %s from new\n", info.ToStringAddrPort());
-                return {info, info.m_last_try};
+        search_tried = insecure_rand.randbool();
+    }
+
+    const int bucket_count{search_tried ? ADDRMAN_TRIED_BUCKET_COUNT : ADDRMAN_NEW_BUCKET_COUNT};
+
+    //  Loop through the addrman table until we find an appropriate entry
+    double chance_factor = 1.0;
+    while (1) {
+        // Pick a bucket, and an initial position in that bucket.
+        int bucket = insecure_rand.randrange(bucket_count);
+        int initial_position = insecure_rand.randrange(ADDRMAN_BUCKET_SIZE);
+
+        // Iterate over the positions of that bucket, starting at the initial one,
+        // and looping around.
+        int i;
+        for (i = 0; i < ADDRMAN_BUCKET_SIZE; ++i) {
+            int position = (initial_position + i) % ADDRMAN_BUCKET_SIZE;
+            int node_id = LookupAddrmanEntry(search_tried, bucket, position);
+            if (node_id != -1) {
+                if (network) {
+                    const auto it{mapInfo.find(node_id)};
+                    const auto info{it->second};
+                    if (info.GetNetwork() == *network) break;
+                } else {
+                    break;
+                }
             }
-            // Otherwise start over with a (likely) different bucket, and increased chance factor.
-            fChanceFactor *= 1.2;
         }
+
+        // If the bucket is entirely empty, start over with a (likely) different one.
+        if (i == ADDRMAN_BUCKET_SIZE) continue;
+
+        // Find the entry to return.
+        int position = (initial_position + i) % ADDRMAN_BUCKET_SIZE;
+        int nId = LookupAddrmanEntry(search_tried, bucket, position);
+        const auto it_found{mapInfo.find(nId)};
+        assert(it_found != mapInfo.end());
+        const AddrInfo& info{it_found->second};
+
+        // With probability GetChance() * chance_factor, return the entry.
+        if (insecure_rand.randbits(30) < chance_factor * info.GetChance() * (1 << 30)) {
+            std::string table_name;
+            search_tried ? table_name = "tried" : table_name = "new";
+            LogPrint(BCLog::ADDRMAN, "Selected %s from %s\n", info.ToStringAddrPort(), table_name);
+            return {info, info.m_last_try};
+        }
+
+        // Otherwise start over with a (likely) different bucket, and increased chance factor.
+        chance_factor *= 1.2;
+    }
+}
+
+int AddrManImpl::LookupAddrmanEntry(bool use_tried, int bucket, int position) const
+{
+    AssertLockHeld(cs);
+
+    if (use_tried) {
+        return vvTried[bucket][position];
+    } else {
+        return vvNew[bucket][position];
     }
```

Comment:
Yes, but I am not saying to do it (use `std::array` and `.at()`). It is just an option, in case you have not considered it. Up to you.

Btw, a difference between an `assert()` and `.at()` is that the former will call `abort()` and will inevitably cause the program to exit. OTOH `.at()` will throw `std::out_of_range` which the caller can catch and continue the execution. I don't have a strong opinion which one is more preferable to use here. Maybe either one is ok. Usually on such "programmer error" we want to really stop the whole program, OTOH `.at()` is much more elegant to write :) and if the exception is unhandled then it will also cause the program to exit.